File: src/routes/product.routes.ts
import { Hono } from "hono";
import { type Env } from "@/types/hono";
import { ProductService } from "@/services/product.service";
import { successResponse } from "@/utils/response";
import { authMiddleware } from "@/middleware/auth";
import { z } from "zod";
import { currencySchema } from "@/types/currency";
import { isEmpty } from "es-toolkit/compat";

export const productRoutes = new Hono<Env>();

let service: ProductService | null = null;

function getService(): ProductService {
  if (!service) {
    service = new ProductService();
  }
  return service;
}

// Apply auth middleware
productRoutes.use("*", authMiddleware);

// Input schemas
const createProductSchema = z.object({
  name: z.string().min(2, "En az 2 karakter gereklidir.").max(100),
  code: z.string().min(2, "En az 2 karakter gereklidir.").max(100),
  price: z.coerce.number().positive("Pozitif bir fiyat girilmelidir"),
  currency: currencySchema,
  vendorId: z
    .string()
    .optional()
    .transform((val) => (val && val.trim() !== "" ? val : undefined)),
  description: z
    .string()
    .max(1000)
    .optional()
    .transform((val) => (val && val.trim() !== "" ? val : undefined)),
});

const updateProductSchema = createProductSchema.partial();

productRoutes.get("/list/all", async (c) => {
  const user = c.get("user");
  if (isEmpty(user.companyId)) {
    return c.json(successResponse([]));
  }

  const products = await getService().listAllProductsForCompany(user.companyId!);

  return c.json(successResponse(products));
});

// GET /api/products - List products for user's company (from company DB, with vendor & priceWithRate)
productRoutes.get("/", async (c) => {
  const user = c.get("user");

  const products = await getService().listProductsForCompany(user.companyId!);

  return c.json(successResponse(products));
});

// GET /api/products/vendor/:vendorId - List products by vendor (with vendor & priceWithRate)
productRoutes.get("/vendor/:vendorId", async (c) => {
  const user = c.get("user");
  const vendorId = c.req.param("vendorId");

  const products = await getService().listProductsByVendor(user.companyId!, vendorId);

  return c.json(successResponse(products));
});

// GET /api/products/:id - Get product detail (with vendor & priceWithRate)
productRoutes.get("/:id", async (c) => {
  const user = c.get("user");
  const id = c.req.param("id");

  const product = await getService().getProduct(user.companyId!, id);

  return c.json(successResponse(product));
});

// POST /api/products - Create product in company database
productRoutes.post("/", async (c) => {
  const user = c.get("user");
  const body = await c.req.json();
  const input = createProductSchema.parse(body);

  const product = await getService().createProduct(user.companyId!, input);

  return c.json(successResponse(product), 201);
});

// PUT /api/products/:id - Update product in company database
productRoutes.put("/:id", async (c) => {
  const user = c.get("user");
  const id = c.req.param("id");
  const body = await c.req.json();
  const input = updateProductSchema.parse(body);

  const product = await getService().updateProduct(user.companyId!, id, input);

  return c.json(successResponse(product));
});

// DELETE /api/products/:id - Delete product from company database
productRoutes.delete("/:id", async (c) => {
  const user = c.get("user");
  const id = c.req.param("id");

  await getService().deleteProduct(user.companyId!, id);

  return c.json(successResponse({ message: "Product deleted successfully" }));
});

File: fisoft/src/main/kotlin/routes/ProductRoutes.kt
package com.ilyasdemirkiran.routes

import com.ilyasdemirkiran.middleware.authenticateUser
import com.ilyasdemirkiran.middleware.currentCompanyId
import com.ilyasdemirkiran.middleware.currentUser
import com.ilyasdemirkiran.services.ProductService
import com.ilyasdemirkiran.types.Money
import com.ilyasdemirkiran.utils.successResponse
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.serialization.Serializable

@Serializable
data class CreateProductReq(
    val name: String,
    val code: String,
    val price: Money,
    val currency: String = "TRY",
    val vendorId: String,
    val description: String? = null,
    val imageUrl: String? = null
)

fun Route.productRoutes() {
    val productService = ProductService()

    route("/api/products") {
        get {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val products = productService.listProductsForCompany(companyId)
            call.respond(HttpStatusCode.OK, successResponse(products))
        }

        get("/list/all") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val products = productService.listProductsForCompany(companyId)
            call.respond(HttpStatusCode.OK, successResponse(products))
        }

        get("/vendor/{vendorId}") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val vendorId = call.parameters["vendorId"] ?: ""
            val products = productService.listProductsByVendor(vendorId, companyId)
            call.respond(HttpStatusCode.OK, successResponse(products))
        }

        post {
            call.authenticateUser()
            val user = call.currentUser
            val req = call.receive<CreateProductReq>()
            val product = productService.createProduct(user.role, req.name, req.code, req.price, req.currency, req.vendorId, req.description, req.imageUrl)
            call.respond(HttpStatusCode.Created, successResponse(product))
        }

        get("/{id}") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val id = call.parameters["id"] ?: ""
            val product = productService.getProduct(id, companyId)
            call.respond(HttpStatusCode.OK, successResponse(product))
        }

        put("/{id}") {
            call.authenticateUser()
            val user = call.currentUser
            val id = call.parameters["id"] ?: ""
            val req = call.receive<CreateProductReq>()
            val updated = productService.updateProduct(user.role, id, req.name, req.code, req.price, req.currency, req.description, req.imageUrl)
            call.respond(HttpStatusCode.OK, successResponse(updated))
        }

        delete("/{id}") {
            call.authenticateUser()
            val user = call.currentUser
            val id = call.parameters["id"] ?: ""
            productService.deleteProduct(user.role, id)
            call.respond(HttpStatusCode.OK, successResponse(mapOf("success" to true)))
        }
    }
}
